src/libos/src/net/socket/unix/stream/endpoint.rs

```diff
@@ -108,3 +147,38 @@ impl Inner {
 // TODO: Add SO_SNDBUF and SO_RCVBUF to set/getsockopt to dynamcally change the size.
 // This value is got from /proc/sys/net/core/rmem_max and wmem_max that are same on linux.
 pub const DEFAULT_BUF_SIZE: usize = 208 * 1024;
+
+/// An observer used to observe both reader and writer of the endpoint. It also contains a
+/// notifier that relays the notification of the endpoint.
+pub(super) struct RelayNotifier {
+    notifier: IoNotifier,
+}
+
+impl RelayNotifier {
+    pub fn new() -> Self {
+        let notifier = IoNotifier::new();
+        Self { notifier }
+    }
+
+    pub fn notifier(&self) -> &IoNotifier {
+        &self.notifier
+    }
+}
+
+impl Observer<IoEvents> for RelayNotifier {
+    fn on_event(&self, event: &IoEvents, _metadata: &Option<Weak<dyn Any + Send + Sync>>) {
+        let mut event = event.clone();
+        // The event of pipe should not be broadcasted directly to socket.
+        // The event transform should be consistant with poll.
+        if event.contains(IoEvents::HUP) {
+            event -= IoEvents::HUP;
+            event |= IoEvents::RDHUP;
+            if event.contains(IoEvents::ERR) {
+                event -= IoEvents::ERR;
+                event |= IoEvents::HUP;
+            }
+        }
```

Comment:
Need to confirm that the resulting event is consistent with the semantics of Linux.